vars/tmsUpload.groovy

```diff
@@ -127,13 +138,85 @@ void call(Map parameters = [:]) {
             }
 
             def token = tms.authentication(uaaUrl, clientId, clientSecret)
+            Map nodeIdExtDesMap = [:]
+            
+            // validate the whole mapping and then throw errors together, 
+            // so that user can get them in one pipeline run
+            // put the validation here, because we need uri and token to call tms get nodes api
+            if(nodeExtDescriptorMapping) {
+                def errorPathList = []
+                def errorMtaId = []
+                def errorNodeNameList = []
+                def errorMsg = ""
+                def mtaYaml
+                List nodes = tms.getNodes(uri, token).getAt("nodes");
+                
+                // get mta.yml to validate mta version and mta id
+                if(fileExists("mta.yaml")) {
+                    mtaYaml = readYaml file: "mta.yaml"
+                    if (!mtaYaml.ID) {
+                        error "Property 'ID' is not found in mta.yaml."
+                    }
+                } else {
+                    error "mta.yaml is not found in the root folder of the project."
+                }
+                
+                if(mtaVersion != "*" && mtaVersion != mtaYaml.version) {
+                    errorMsg = "Parameter 'mtaVersion' dose not match the MTA version in mta.yaml. "
+                }
+                
+                nodeExtDescriptorMapping.each{ key, value ->
+                    if(nodes.any {it.name == key}) {
+                        nodeIdExtDesMap.put(nodes.find {it.name == key}.getAt("id"), value)
+                    } else {
+                        errorNodeNameList.add(key)
+                    }
+                    
+                    if(!fileExists(value)) {
+                        errorPathList.add(value)
+                    } else {
+                        if(mtaYaml.ID != getMtaId("${value}")) {
+                            errorMtaId.add(value)
+                        }
+                    }
+                }
+                
+                if(!errorPathList.isEmpty() || !errorMtaId.isEmpty() || !errorNodeNameList.isEmpty() ) {
+                    if(!errorPathList.isEmpty()) {
+                        errorMsg += "MTA extension descriptor files ${errorPathList} don't exist. "
+                    }
+                    if(!errorMtaId.isEmpty()) {
+                        errorMsg += "MTA ID in MTA extension descriptor files ${errorMtaId} is incorrect. "
+                    }
+                    if(!errorNodeNameList.isEmpty()) {
+                        errorMsg += "Nodes ${errorNodeNameList} don't exist. Please check the node name or creat these nodes."
+                    }
+                    error(errorMsg)
+                }
+            }
+            
+            if(nodeIdExtDesMap) {
+                nodeIdExtDesMap.each{ key, value ->
```

Comment:
> The code of uploading of extenion descriptors might be moved into the `if(nodeExtDescriptorMapping) {...}`, since it is relevant only for that case.

I may prefer to keeping it separating from the validation part.

Reply:
I imagine it in the following way: inside of the if-condition that checks for existence of mapping between node names and extension descriptors do the further two thins: 1. execute validation and form another map between node ids and extension descriptors 2. execute uploading

Reply:
Hi Artem, I do not have a strong preference. I modified codes as you suggested in this and previous comments in my next commit.